sync/category.go

```diff
@@ -0,0 +1,93 @@
+//    * If not found, it will recreate the model.InvoiceCategory.
+//    * If found and model.InvoiceCategory.Name is up-to-date, it will return without error
+//    * If found and model.InvoiceCategory.Name differs from db.Category.Source, the model.InvoiceCategory is updated/reset.
+func (s *InvoiceCategoryReconciler) Reconcile(ctx context.Context, category *db.Category) error {
+	converted := model.InvoiceCategory{}
+	err := CategoryConverter{}.ToInvoiceCategory(category, &converted)
+	if err != nil {
+		return err
+	}
+	if !category.Target.Valid {
+		return s.findOrCreateCategory(ctx, category, converted)
+	}
+	return s.updateCategoryIfNeeded(ctx, category, converted)
+}
+
+func (s *InvoiceCategoryReconciler) findOrCreateCategory(ctx context.Context, category *db.Category, converted model.InvoiceCategory) error {
+	found, err := s.odoo.SearchInvoiceCategoriesByName(ctx, converted.Name)
+	if err != nil {
+		return err
+	}
+	if len(found) == 0 {
+		return s.createCategoryInOdoo(ctx, category)
+	}
+	category.Target.String, category.Target.Valid = strconv.Itoa(found[0].ID), true
+	return nil
+}
+
+func (s *InvoiceCategoryReconciler) createCategoryInOdoo(ctx context.Context, category *db.Category) error {
+	ic := newInvoiceCategory(nil)
+	err := CategoryConverter{}.ToInvoiceCategory(category, &ic)
+	if err != nil {
+		return err
+	}
+	created, err := s.odoo.CreateInvoiceCategory(ctx, ic)
+	if err != nil {
+		return err
+	}
+	category.Target.String, category.Target.Valid = strconv.Itoa(created.ID), true
+	return nil
+}
+
+func (s *InvoiceCategoryReconciler) updateCategoryIfNeeded(ctx context.Context, category *db.Category, ic model.InvoiceCategory) error {
+	// TODO: re Idempotency, should we search by name instead? Maybe the category in odoo got deleted & recreated with a new ID manually
+	existingIC, err := s.odoo.FetchInvoiceCategoryByID(ctx, ic.ID)
+	if err != nil {
+		return err
+	}
+	if existingIC == nil {
+		// The category in Odoo might have been deleted, let's recreate
+		return s.createCategoryInOdoo(ctx, category)
+	}
+
+	cv := CategoryConverter{}
+	if !cv.IsSame(*existingIC, category) {
+		// Updating existing category should rarely be the case.
+		// Possible case is given if the category name has been manually updated in Odoo, in that case revert it since the DB is authoritative.
+		return s.updateCategoryInOdoo(ctx, category, existingIC)
+	}
+	return nil
+}
+
+func (s *InvoiceCategoryReconciler) updateCategoryInOdoo(ctx context.Context, category *db.Category, existing *model.InvoiceCategory) error {
+	ic := newInvoiceCategory(existing)
+	err := CategoryConverter{}.ToInvoiceCategory(category, &ic)
+	if err != nil {
+		return err
+	}
+	return s.odoo.UpdateInvoiceCategory(ctx, ic)
+}
+
+func newInvoiceCategory(existing *model.InvoiceCategory) model.InvoiceCategory {
+	ic := model.InvoiceCategory{}
+	if existing != nil {
+		ic = *existing
+	}
```

Comment:
This check IMO should not be the job of this function

Reply:
I agree, I'm not quite happy with that myself :)